viewing_party/party.py

```diff
@@ -1,23 +1,155 @@
+    if not user_data["watched"]:
+        return 0
+
+    ratings_sum = sum([movie["rating"] for movie in user_data["watched"]])
+    movie_count = len(user_data["watched"])
+
+    return ratings_sum/movie_count
+
+
+def get_most_watched_genre(user_data):
+    if not user_data["watched"]:
+        return None
+
+    genre_count = {}
+
+    for movie in user_data["watched"]:
+        if movie["genre"] in genre_count:
+            genre_count[movie["genre"]] += 1
+        else:
+            genre_count[movie["genre"]] = 1
+
+    return max(genre_count, key=genre_count.get)
+
+
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+
+def get_unique_watched(user_data):
+    friends_watched = []
+    list_of_movies = []
+
+    for i in user_data["friends"]:
+        for j in i["watched"]:
+            friends_watched.append(j["title"])
+
+    for i in user_data["watched"]:
+        if i["title"] not in set(friends_watched):
+            list_of_movies.append(i)
+
+    return list_of_movies
+
+
+def get_friends_unique_watched(user_data):
+    user_titles = []
+    for movie in user_data["watched"]:
+        user_titles.append(movie["title"])
+
+    friends_titles = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_titles.append(movie["title"])
+
+    friends_unique_titles = []
+    for item in set(friends_titles):
+        if item not in set(user_titles):
+            friends_unique_titles.append(item)
+
+    friends_unique_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            if movie["title"] in set(friends_unique_titles):
+                friends_unique_movies.append(movie)
+
+    friends_unique_movies_no_duplicates = []
+    for movie in friends_unique_movies:
+        if movie not in friends_unique_movies_no_duplicates:
+            friends_unique_movies_no_duplicates.append(movie)
+
+    return friends_unique_movies_no_duplicates
+
+
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+
+def get_available_recs(user_data):
+    movies_to_consider = get_friends_unique_watched(user_data)
+    subscriptions_owned = user_data["subscriptions"]
+
+    return [movie for movie in movies_to_consider
+            if movie["host"] in subscriptions_owned]
+
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
 
+
+def get_new_rec_by_genre(user_data):
+    movies_to_consider = get_friends_unique_watched(user_data)
+
+    user_watched_genres = []
+    for k in user_data["watched"]:
+        user_watched_genres.append(k["genre"])
+
+    if len(user_watched_genres) == 0:
+        return []
+
+    most_frequent_genre = max(set(user_watched_genres), key=user_watched_genres.count)

```

Comment:
Instead of rewriting similar logic, you could have leveraged the function `get_most_watched_genre` above and refactored much of this. It's generally preferable to reuse code whenever possible!